File: c/abb.c
#include "abb.h"
#include <stdio.h>
#include <stdlib.h>

void  error ( char mensaje_error[]) {

	printf("%s\n",mensaje_error);
}// Fin error

void crearNodoA ( tPosA *p) {
	
	*p = malloc(sizeof(tPosA));
	if (*p==NULL) {
		error(" *** abb.crearNodoA: no hay memoria");
	}
} //Fin crearNodoA

/**************************************************************/

void arbolVacio (tABB *arbol){
	*arbol = NULL;
}//Fin arbolVacio

/************************Insertar_Nodo*************************/
//insertar recursivo
void insertar_r (tABB * arbol, tClave clave){
	tABB nuevo;
	if (*arbol==NULL){
		crearNodoA(arbol);
		(*arbol)->clave = clave;
		(*arbol)->izdo = NULL;
		(*arbol)->dcho = NULL;
	}else 
		if (clave < (*arbol)->clave){
			insertar_r(&(*arbol)->izdo,clave);
		}else 
			if (clave > (*arbol)->clave) {
				insertar_r(&(*arbol)->dcho, clave);			
			}//Se ignoran los dublicados*/
}//fin insertar_r

//insertar iterativo
void insertar_i (tABB * arbol, tClave clave){
	//Variables
	tABB nuevo, padre, hijo;
	//Fin Variables
		crearNodoA(&nuevo);
		nuevo->clave = clave;
		nuevo->izdo = NULL;
		nuevo->dcho = NULL;
		if (*arbol==NULL) {
			//Caso arbol vacio
			*arbol = nuevo;
		}else{
			//Caso arbol no vacio
			padre = NULL;
			hijo = *arbol;
			while ((hijo != NULL) && (hijo->clave != clave)) {
				padre = hijo;
				if (clave < hijo->clave) {
					hijo = hijo->izdo;
				}else {
					hijo = hijo->dcho;
				}		
			} 
			if (hijo == NULL) {
				if (clave < padre->clave) {
					padre->izdo = nuevo;
				}else {
					padre->dcho = nuevo;			
				}// Se ignoran los duplicados			
			}
		}
}// Fin insertar iterativo

void insertarClave (tABB *arbol, tClave clave) {
	insertar_r(arbol,clave);
}

/*********************buscar_nodo**************************/
//buscar recursivo
tPosA buscar_r(tABB arbol, tClave clave){
	if (arbol==NULL) {
		return NULL;		
	}else 
		if (arbol->clave == clave) {
			return arbol;	
		}else 
			if (clave < arbol->clave) {
				buscar_r(arbol->izdo, clave);	
			} else buscar_r(arbol->dcho,clave);
}//Fin Buscar_R 

//buscar iterativo
tPosA buscar_i (tABB *arbol, tClave clave) {
	//Variables
	tPosA nodo;
	//Fin variables

	nodo = *arbol;
	while ((nodo != NULL) && (nodo->clave != clave)){
		if (clave < nodo->clave){
			nodo = nodo->izdo;
		}else{
			nodo = nodo->dcho;
		}
	}
	return nodo;
}//Fin buscar_i

tABB buscarClave (tABB arbol, tClave clave) {
	return buscar_r(arbol,clave);
}

/**************************************************************/

tABB hijoIzquierdo(tABB arbol) {	
		return arbol->izdo;
}//Fin hijoIzquierdo

tABB hijoDerecho(tABB arbol) {	
		return arbol->dcho;
}//Fin hijoDerecho

tClave Raiz(tABB arbol) {
	return arbol->clave;	
}//Fin Raiz

bool esArbolVacio(tABB arbol) {
	return (arbol==NULL);
}//Fin esArbolVacio

/**************************ELiminar_Nodo**********************/
void eliminar_r (tABB * arbol, tClave clave) {
	//Variables
	tABB aux;
	//Fin variables
	
	void sup2 (tABB *arbol2){
		if ((*arbol2)->dcho != NULL){
			sup2(&(*arbol2)->dcho);
		}else {
			aux->clave = (*arbol2)->clave;
			aux = *arbol2;
			*arbol2= (*arbol2)->izdo;
		}
	
	}//final sup2

	if (arbol != NULL) {
		if (clave < (*arbol)->clave) {
			eliminar_r(&(*arbol)->izdo,clave);
		}else 
			if (clave > (*arbol)->clave) {
				eliminar_r(&(*arbol)->dcho,clave);
			}else {
				aux = *arbol;
				if ((*arbol)->izdo==NULL){
					*arbol = (*arbol)->dcho;			
				} else 
					if ((*arbol)->dcho==NULL){
						*arbol = (*arbol)->izdo;
					}else sup2(&(*arbol)->izdo);
				free(aux);
			} //else
	}//if
		
}//fin eliminar recursivo

//eliminar iterativo
void eliminar_i (tABB *arbol, tClave clave) {
	//Variables
	int numHijos;
	tABB sup, //nodo a suprimir
		 pSup, //padre del nodo a suprimir
		 hijoNoVacio, //si el nodo a suprimir solo tiene un hijo, el no vacio
		 sucIzMax; //si el nodo a suprimir tiene dos hijos, el nodo con mayor valor del subarbol izquierdo
	//Fin Variables
	
	//Buscar el nodo a suprimir
	pSup = NULL;	
	sup = *arbol;
	while ((sup != NULL) && (sup->clave != clave)) {
		pSup = sup;
		if (clave < sup->clave) {
			sup = sup->izdo;
		} else {
			sup = sup->dcho;		
		}
	}//while
	
	// si la clave no pertenece al arbol no se hace nada
	if (sup != NULL){
		//Cuenta el numero de hijos
		numHijos = 0;
		if (sup->izdo != NULL){
			numHijos++;
		}
		if (sup->dcho != NULL) {
			numHijos++;		
		}

		switch (numHijos) {
			//suprimir nodo hoja
			case 0 :
				if (pSup == NULL) {
					arbol = NULL; //la raiz era el unico nodo del arbol
				}else if (pSup->izdo == sup){ 
					pSup->izdo = NULL;				
				}else pSup->dcho = NULL; 
				break;
			
			//Suprimir nodo con un solo hijo
			case 1:
				if (sup->izdo == NULL) {
					hijoNoVacio = sup->dcho;				
				}else {
					hijoNoVacio = sup->izdo;			
				}
				
				if (pSup == NULL) {
					*arbol = hijoNoVacio; //borramos la raiz
				}else if (pSup->izdo == sup) {
					pSup->izdo = hijoNoVacio;
				}else
					pSup->dcho = hijoNoVacio;
				break;
			//suprimir nodo con dos hijos
			case 2:	
				pSup = sup;
				sucIzMax = sup->izdo;
				while (sucIzMax->dcho != NULL) {
					pSup = sucIzMax;
					sucIzMax = sucIzMax->dcho;
				}
				
				sup->clave = sucIzMax->clave;
				if (pSup == sup) {
					pSup->izdo = sucIzMax->izdo;
				}else 
					pSup->dcho = sucIzMax->izdo;
				
				sup = sucIzMax;	
				break;					
		}//switch
		
		free(sup);	
	}//if
}// eliminar_i


void eliminarClave ( tABB *arbol, tClave clave) {
	eliminar_r(arbol,clave);
}//eliminar clave

File: OCaml/abb.ml
type abb = Null | Node of (int * abb ref * abb ref);;

exception Invalid_arg;;

let arbolVacio a =
	a := Null;;

let crearNodo p c = 
	p := Node (c, ref Null, ref Null);;

let hijoIzquierdo = function
	| Null -> raise (Invalid_arg)
	| Node (_,i,_) -> i;;

let hijoDerecho = function
	| Null -> raise (Invalid_arg)
	| Node (_,_,d) -> d;;

let raiz = function
	| Null -> raise (Invalid_arg)
	| Node (r,_,_) -> r;;

let esArbolVacio = function
	| Null -> true
	| Node _ -> false;;

let rec insertar_r a c =
	if !a = Null then
		crearNodo a c
	else if c < (raiz !a) then
		insertar_r (hijoIzquierdo !a) c
	else if c > (raiz !a) then
		insertar_r (hijoDerecho !a) c
	else ();;

let insertar_i a c =
	let nuevo = ref Null in
	let padre = ref Null in
	let hijo = ref Null in
	crearNodo nuevo c;
	if !a = Null then a := !nuevo
	else begin
		hijo := !a;
		while (!hijo <> Null) && (raiz !hijo <> c) do
			padre := !hijo;
			if c < (raiz !hijo)
			then hijo := !(hijoIzquierdo !hijo)
			else hijo := !(hijoDerecho !hijo)
		done;

		if !hijo = Null then
			if c < (raiz !padre) then
				(hijoIzquierdo !padre) := !nuevo
			else
				(hijoDerecho !padre) := !nuevo
	end;;

let rec buscar_r a c =
	if !a = Null then
		ref Null
	else if c = (raiz !a) then a
	else if c < (raiz !a) then
		let izq = (hijoIzquierdo !a) in
		buscar_r izq c
	else
		let der = (hijoDerecho !a) in
		buscar_r der c;;

let buscar_i a c =
	let nodo = ref Null in
	nodo := !a;
	while !nodo <> Null && (raiz !nodo) <> c do
		if c < (raiz !nodo)
		then nodo := !(hijoIzquierdo !nodo)
		else nodo := !(hijoDerecho !nodo)
	done;
	nodo;;

let rec eliminar_r a c =
	let aux = ref Null in
	let rec sup2 b =
		let Node (clave,izquierdo,derecho) = !b in
		if !derecho <> Null then
			sup2 derecho
		else
			let Node (_,auxizq,auxder) = !aux in (
				aux := Node (clave,auxizq,auxder);
				b := !izquierdo )
	in if !a <> Null then
		let Node (ra,izq,der) = !a in
		if c < ra then
			eliminar_r izq c
		else if c > ra then
			eliminar_r der c
		else (
			aux := !a;
			if !izq = Null then a := !der
			else if !der = Null then a := !izq
			else (
				sup2 (hijoIzquierdo !a);
				a := !aux ))
	else ();;

let eliminar_i a c =
	let sup = ref Null in
	let pSup = ref Null in
	let hijoNoVacio = ref Null in
	let sucIzMax = ref Null in
	sup := !a;
	while !sup <> Null && (raiz !sup) <> c do
		pSup := !sup;
		if c < (raiz !sup) then sup := !(hijoIzquierdo !sup)
		else sup := !(hijoDerecho !sup)
	done;

	if !sup <> Null then
		match !sup with
			| Node (_, i, d) when !i = Null && !d = Null ->
				if !pSup = Null then a := Null
				else if sup = (hijoIzquierdo !pSup) then
					(hijoIzquierdo !pSup) := Null
				else (hijoDerecho !pSup) := Null
			| Node (_,i,d) when (!i = Null) <> (!d = Null) ->
				hijoNoVacio := if !i = Null then !d else !i;
				if !pSup = Null then a := !hijoNoVacio
				else if (hijoIzquierdo !pSup) = sup then
					(hijoIzquierdo !pSup) := !hijoNoVacio
				else (hijoDerecho !pSup) := !hijoNoVacio
			| Node _ ->
				pSup := !sup;
				sucIzMax := !(hijoIzquierdo !sup);
				while !(hijoDerecho !sucIzMax) <> Null do
					pSup := !sucIzMax;
					sucIzMax := !(hijoDerecho !sucIzMax)
				done;
				
				let Node(_,izq,der) = !sup in
				let Node(_,pi,pd) = !pSup in (
				sup := Node (raiz !sucIzMax, izq, der);
				if pSup = sup then
					pi := !(hijoIzquierdo !sucIzMax)
				else pd := !(hijoDerecho !sucIzMax);
				sup := !sucIzMax )

let p = ref Null;;
let test () =
	insertar_i p 10;
	insertar_i p 5;
	insertar_i p 6;
	eliminar_i p 5;;
